src/Helper/HeaderHelper.php

```diff
@@ -0,0 +1,128 @@
+<?php
+
+namespace Yiisoft\Yii\Web\Helper;
+
+use Psr\Http\Message\RequestInterface;
+
+final class HeaderHelper
+{
+    /**
+     * Explode header value to value and parameters (eg. text/html;q=2;version=6)
+     * @param string $headerValue
+     * @return array first element is the value, and key-value are the parameters
+     */
+    public static function getValueAndParameters(string $headerValue): array
+    {
+        $headerValue = trim($headerValue);
+        if (strlen($headerValue) === 0) {
+            return [];
+        }
+        $parts = preg_split('/\s*;\s*/', $headerValue, -1, PREG_SPLIT_NO_EMPTY);
+        $output = [array_shift($parts)];
+        foreach ($parts as $part) {
+            [$key, $headerValue] = explode('=', $part, 2);
+            $output[$key] = $headerValue;
+        }
+        return $output;
+    }
+
+    /**
+     * Getting header value as q factor sorted list
+     * @param string|string[] $values Header value as a comma-separated string or already exploded string array.
+     * @see getValueAndParameters
+     * @link https://developer.mozilla.org/en-US/docs/Glossary/Quality_values
+     */
+    public static function getByQFactorSortedList($values, bool $outputValuesOnly = false): array
+    {
+        if (is_string($values)) {
+            $values = preg_split('/\s*,\s*/', trim($values), -1, PREG_SPLIT_NO_EMPTY);
+        }
+        if (!is_array($values)) {
+            throw new \InvalidArgumentException('Values ​​are neither array nor string!');
+        }
+        if (count($values) === 0) {
+            return [];
+        }
+        $output = [];
+        foreach ($values as $value) {
+            $parse = self::getValueAndParameters($value);
+            $q = $parse['q'] ?? 1.0;
+            if (is_string($q) && preg_match('/^(?:0(?:\.\d{1,3})?|1(?:\.0{1,3})?)$/', $q) === 0) {
+                throw new \InvalidArgumentException('Invalid q factor!');
+            }
+            $parse['q'] = floatval($q);
+            $output[] = $parse;
+        }
+        usort($output, function ($a, $b) {
+            $a = $a['q'];
+            $b = $b['q'];
+            if ($a === $b) {
+                return 0;
+            }
+            return $a > $b ? -1 : 1;
+        });
+        if (!$outputValuesOnly) {
+            return $output;
+        }
+        foreach ($output as &$value) {
+            $value = array_shift($value);
+        }
+        return $output;
+    }
+
+    /**
+     * @param $values string|string[]|RequestInterface $values Header value as a comma-separated string
```

Comment:
It is better to have at least a separate method that works with RequestInterface. Having too many types in a single parameter adds to overall method complexity including its usage.

Reply:
How about using only `RequestInterface` as input?